doc/modules/linear_model.rst

```diff
@@ -842,6 +848,64 @@ Ridge Regression`_, see the example below.
 
     .. [4] Tristan Fletcher: `Relevance Vector Machines Explained <https://citeseerx.ist.psu.edu/doc_view/pid/3dc9d625404fdfef6eaccc3babddefe4c176abd4>`_
 
+.. _linear_regression_intercept:
+
+Note on the handling of the intercept in linear regression models
+-----------------------------------------------------------------
+
+Recall the definition of the ridge estimator:
+
+.. math::  \min_{w, w_0} { ||X w + w_0 1_s - y||_2 ^ 2 + \alpha ||w||_2 ^ 2}
+
+Here :math:`1_s` is a vector of size `n_samples` where all entries are 1.
+:math:`w_0` is a scalar parameter and is not regularized.
+
+Let's note :math:`\bar{X}` the column vector whose entries the means of the
+columns of :math:`X` and :math:`\bar{y}` the mean scalar value of :math:`y`,
+:math:`X_c` the centered version of :math:`X`, :math:`y_c` the centered version
+of :math:`y` such that:
+
+.. math::  X = X_c + 1_s \bar{X}^{T}  \text{ and } y = y_c + \bar{y} 1_s
+
+Optimizing the original ridge objective with intercept is equivalent to solving
+the following ridge problem on the centered dataset with no intercept:
+
+.. math:: \hat{w} = \min_{w} { ||X_c w - y_c||_2 ^ 2 + \alpha ||w||_2 ^ 2}
+
+and subsequently define the intercept as :math:`\hat{w_0} = \bar{y} -
+\bar{X}^{T} \hat{w}`.
+
+To see this, let's rewrite the data-fit term of the original ridge problem as
+follows (after rearranging the terms and factorizing by :math:`1_s`):
+
+.. math::  ||X w + w_0 1_s - y||_2^2 = ||X_c w - y_c +  (\bar{X}^{T} w  + w_0  - \bar{y}) 1_s ||_2^2
+
+Let's define the scalar :math:`\bar{r} = \bar{X}^{T} w  + w_0  - \bar{y}`. We
+can now expand:
+
+.. math::  ||X w + w_0 1_s - y||_2^2 = ||X_c w - y_c ||_2^2 + ||\bar{r} 1_s ||_2^2 + 2 \bar{r} 1_s^{T} (X_c w - y_c)
+
+If :math:`n` is the number of samples, we have:
+
+.. math::  ||X w + w_0 1_s - y||_2^2 = ||X_c w - y_c ||_2^2 + \bar{r}^2 n + 2 \bar{r} (1_s^{T} X_c w - 1_s^{T} y_c)
+
+Since :math:`X_c` and :math:`y_c` are the centered versions of X and y, we have
+:math:`1_s^{T} X_c = 0` and :math:`1_s^{T} y_c = 0`. Therefore:
+
+.. math::  ||X w + w_0 1_s - y||_2^2 + \alpha ||w||_2^2 = ||X_c w - y_c ||_2^2  + \alpha ||w||_2^2 + \bar{r}^2 n
+
+The first two terms of the right hand side do not depend on :math:`w_0` and are
+minimized by :math:`\hat{w}` the solution of the centered ridge without
+intercept. The third term is positive and therefore subsequently minimized to
+zero by setting :math:`\hat{w_0} = \bar{y} - \bar{X}^{T} \hat{w}`.
+
+Note that the same argument holds for the OLS estimator (for the specific case
+where :math:`\alpha = 0`) or for any other penalized linear regression
+estimator.
+
+Furthermore least-norm solution for the centered OLS problem without intercept
+is also the least-norm solution for non-centered OLS problem with intercept
```

Comment:
This is actually not that obvious in retrospect. When `X.T @ X` is singular, the OLS problem needs to be written as the constrained minimization problem `argmin_{w, w_0} ||w|| st. X w + w_0 = y` in order to admit a unique solution. To show the equivalence with the minimum norm solution to `X_c w = y_c` we would first need to show that for any solutions (w^*, w_0^*) to the `X w + w_0 = y`  problem then `w^*` is also a solutions to `X_c w = y_c`.

Reply:
Or maybe we could prove the equivalence by continuity by taking the limit of the 2 ridge solutions when alpha tends towards 0?

Reply:
Actually I know how to cover the OLS cased. I will later push a commit to properly cover this case.

Reply:
OK, do ping me when you think this can be reviewed. :+1:

Reply:
I pushed a new commit to extend the doc to that case. The HTML rendering should soon be ready.